### PR TITLE
dts: gen_defines.py: Generate unquoted versions of string props

### PR DESCRIPTION
**Pull request by ulfalizer**
For each `type: string` property, generate an additional `*_UNQUOTED`
macro that with an unquoted value. This can be useful with token pasting
and the like.

Motivated by https://github.com/zephyrproject-rtos/zephyr/issues/21273,
where there's this:

    variant:
        type: string
        required: true
        enum:
            - "epd1p54"
            - "epd1p54b"
            - "epd1p54c"
            - "epd4p2"
            # ...

Depending on the selection, an identifier like `epd1p56b_init` should be
generated.

The macros below now get generated, where the `*_UNQUOTED` version could
be token-pasted to generate `epd1p56b_init`.

    #define DT_..._VARIANT "epd1p54b"
    #define DT_..._VARIANT_UNQUOTED epd1p54b

Something like `*_ID` could've been used instead of `*_UNQUOTED` too, but
`*_UNQUOTED` is nice in that it's greppable and makes the magic stick out
more.

Supporting commit:

```
Rename str2ident() to to_upper_ident() and add a new to_ident() function
that just replaces non-alphanumeric characters with underscores, without
uppercasing the string.

to_ident() will be used in the next commit.

Also simplify the underscore replacement a bit with a regex, and explain
why '+' is replaced with 'PLUS'.
```

### CODE REVIEW

**Comment by ulfalizer** [comment]
Feels a bit magic, and not sure if overkill. Please comment.

Could do it just for properties with `enum:` too, but this is more general.

**Comment by ulfalizer** [comment]
CC @pabigot

File: scripts/dts/gen_defines.py [hunk 1]
```diff
@@ -221,6 +221,7 @@ def write_props(node):
             out_dev(node, ident, 1 if prop.val else 0)
         elif prop.type == "string":
             out_dev_s(node, ident, prop.val)
+            out_dev(node, ident + "_UNSTRING", to_ident(prop.val))
```
Comment (pabigot) [inline scripts/dts/gen_defines.py]: I think `_TOKEN` would be a better suffix as the desire is to have something that can act as a token in preprocessor manipulation, e.g. `DT_PROPERTY##_cfg`.

File: scripts/dts/gen_defines.py [hunk 2]
```diff
@@ -679,16 +686,22 @@ def write_clocks(node):
...
+    # a string suitable for (part of) an identifier.
+    #
+    # Replaces '+' with 'plus' instead of '_'. This is so that to_upper_ident()
+    # generates e.g. DT_COMPAT_ARM_CORTEX_M0PLUS instead of
+    # DT_COMPAT_ARM_CORTEX_M0_ for 'compatible = "arm,cortex-m0+"'.
+
+    return re.sub(r"\W", "_", s.replace("+", "plus"), flags=re.ASCII)
+
+
+def to_upper_ident(s):
```
Comment (pabigot) [inline scripts/dts/gen_defines.py]: It feels a little weird to have this as a separate function; I immediately wonder how it's different from the more basic `to_ident(s).upper()`.  Saving a character doesn't seem worth it.

**Review by pabigot — COMMENTED** [review]
I think this can be helpful, but I also think the Generated Enums proposal will be more generally useful. 
 I.e. having `DT_COMPAT_PROPERTY_ENUMVAL` generated as a boolean property for the specific enumval that's selected (leaving the unselected ones undefined).  A complete solution to #21273 probably requires both.

File: scripts/dts/gen_defines.py (hunk 1 above)
Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: Strings are tokens too, so think it might be confusing.

File: scripts/dts/gen_defines.py (hunk 2 above)
Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: It's so common that it felt worth it to me. Shaves some bytecode too, though it's not likely to be significant.~

**Comment by ulfalizer** [comment]
> I think this can be helpful, but I also think the Generated Enums proposal will be more generally useful.
> I.e. having `DT_COMPAT_PROPERTY_ENUMVAL` generated as a boolean property for the specific enumval that's selected (leaving the unselected ones undefined). A complete solution to #21273 probably requires both.

Yeah, gonna do a separate PR for that one.

**Comment by ulfalizer** [comment]
PR for testing string enum values in preprocessor: https://github.com/zephyrproject-rtos/zephyr/pull/21316

**Comment by ulfalizer** [comment]
One thing that's slightly iffy about this is that the unquoted identifier might clash with other macros in the code, and e.g. get expanded. Maybe not super likely though, and maybe it could be used deliberately (though it might be confusing).

File: scripts/dts/gen_defines.py (hunk 1 above)
Comment (pabigot) [reply scripts/dts/gen_defines.py]: I don't agree.  "unstring" is more likely to be confusing: it's weird and a term nobody uses.  String literals may be tokens, but they can't participate in token pasting, which is the entire reason why we need something distinct from the macro that provides the string value.

Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: `*_TOKEN` would've been very confusing to me at least.

I like `*_UNSTRING`. Tells you what it does. We kinda make up words as we go more than English though.~

Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: `*_UNQUOTED` might work too, if you'd prefer that.

Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: Changed it to `*_UNQUOTED`. Think I like it better than `*_UNSTRING`.

**Comment by ulfalizer** [comment]
Removed `[RFC]` since nothing will happen while it's there.

**Review by pabigot — CHANGES_REQUESTED** [review]
The first commit would be fine as a cleanup.

Blocking the second until we come to agreement in #21369 about how devicetree paths are to be mapped to symbols that won't conflict.

**Comment by galak** [comment]
Closing this as we are moving to the new generator / dt macro's

**Pull request closed without merging**